ql/src/test/org/apache/hadoop/hive/ql/history/TestHiveScribeImpl.java

```diff
@@ -0,0 +1,150 @@
+    planInfoFormat = String.format("PlanInfo(timeStamp:%s, planDetails:PlanDetails(queryId:%s, stageGraph:GraphInfo(), stageList:[], done:[], started:[]))", planTime, queryId);
+    plan = new QueryPlan();
+    plan.setQueryId(queryId);
+    plan.setQuery(new Query());
+
+    oneMapRedStats = new MapRedStats(1, 1, 10, true, "1");
+    mapRedStats = new HashMap();
+    mapRedKey = "Stage 1";
+    mapRedStats.put(mapRedKey, oneMapRedStats);
+  }
+
+  @Test
+  public void testStartQueryMissingSessionState() {
+    hiveScriber.startQuery(queryString, queryId, null, queryStartTime);
+
+    assertTrue(hiveScriber.getQueryStatsMap().keySet().size() == 0);
+  }
+
+  @Test
+  public void testEntireQueryCycle() throws TException, IOException {
+    // Start query
+    sessionState = Mockito.mock(SessionState.class);
+    when(sessionState.getUserName()).thenReturn(username);
+    when(sessionState.getUserIpAddress()).thenReturn(ipAddress);
+    when(sessionState.getSessionId()).thenReturn(sessionId);
+    when(sessionState.getCurrentDatabase()).thenReturn(dataBase);
+    when(sessionState.getMapRedStats()).thenReturn(mapRedStats);
+
+    twitterScriber = Mockito.mock(TwitterScriber.class);
+    hiveScriber.hiveHistScriber.scriber = twitterScriber;
+
+    hiveScriber.startQuery(queryString, queryId, sessionState, queryStartTime);
+
+    assertTrue(hiveScriber.getQueryStatsMap().keySet().size() == 1);
+    assertTrue(hiveScriber.getQueryStatsMap().keySet().contains(queryId));
+    assertEquals(queryId, hiveScriber.getQueryStatsMap().get(queryId).getQueryId());
+    assertEquals(queryString, hiveScriber.getQueryStatsMap().get(queryId).getQueryString());
+    assertEquals((long) queryStartTime, hiveScriber.getQueryStatsMap().get(queryId).getStartTime());
+
+    // Start Task
+    hiveScriber.startTask(queryId, taskId, taskName, taskStartTime);
+    assertTrue(hiveScriber.getQueryStatsMap().get(queryId).getTaskProgress().size() == 1);
+    assertEquals(taskStart, hiveScriber.getQueryStatsMap().get(queryId).getTaskProgress().get(0).getProgress());
+
+    // Progress Task
+    hiveScriber.progressTask(queryId, taskId, taskProgressTime);
+    assertTrue(hiveScriber.getQueryStatsMap().get(queryId).getTaskProgress().size() == 2);
+    assertEquals(taskProgress, hiveScriber.getQueryStatsMap().get(queryId).getTaskProgress().get(1).getProgress());
+
+    // End Task
+    hiveScriber.endTask(queryId, taskId, taskEndTime);
+    assertTrue(hiveScriber.getQueryStatsMap().get(queryId).getTaskProgress().size() == 3);
+    assertEquals(taskEnd, hiveScriber.getQueryStatsMap().get(queryId).getTaskProgress().get(2).getProgress());
+
+    // Progress Plan
+    hiveScriber.logPlanProgress(plan, planTime);
+
+    // End query
+    hiveScriber.endQuery(queryId, sessionState, queryEndTime);
+
+    // Capture log event at scribe()
+    ArgumentCaptor<HiveQueryCompletionEvent> arg = ArgumentCaptor.forClass(HiveQueryCompletionEvent.class);
```

Comment:
I would prefer capture a string here and then decode string as a `HiveQueryCompletionEvent` and finally compare the fields.

Reply:
Updated Mockito capture to return a String, and then decode it to a thrift object and compare.